src/NuGet.Licenses/Services/LicenseExpressionSplitter.cs

```diff
@@ -0,0 +1,137 @@
+        public List<CompositeLicenseExpressionRun> GetLicenseExpressionRuns(LicenseOperator licenseExpressionRoot)
+        {
+            if (licenseExpressionRoot == null)
+            {
+                throw new ArgumentNullException(nameof(licenseExpressionRoot));
+            }
+
+            var runList = new List<CompositeLicenseExpressionRun>();
+            InOrderTraversal(licenseExpressionRoot, runList);
+            return runList;
+        }
+
+        /// <summary>
+        /// Given the original license expression and list of runs produced by <see cref="GetLicenseExpressionRuns"/>
+        /// produces full split of the expression into list of runs that include all the characters of the original
+        /// license expression (including the parentheses and whitespace).
+        /// </summary>
+        /// <param name="licenseExpression">Original license expression.</param>
+        /// <param name="runs">List of runs produced by <see cref="GetLicenseExpressionRuns"/></param>
+        /// <returns>List of runs including the characters that are lost during expression parsing</returns>
+        public List<CompositeLicenseExpressionRun> SplitFullExpression(string licenseExpression, IReadOnlyCollection<CompositeLicenseExpressionRun> runs)
+        {
+            if (licenseExpression == null)
+            {
+                throw new ArgumentNullException(nameof(licenseExpression));
+            }
+
+            if (runs == null)
+            {
+                throw new ArgumentNullException(nameof(runs));
+            }
+
+            var fullRunList = new List<CompositeLicenseExpressionRun>();
+            var startIndex = 0;
+            foreach (var run in runs)
+            {
+                var currentRunStartIndex = licenseExpression.IndexOf(run.Value, startIndex);
+                if (currentRunStartIndex < 0)
+                {
+                    throw new InvalidOperationException($"Unable to find '{run.Value}' portion of the license expression starting from {startIndex} in '{licenseExpression}'");
+                }
+                if (currentRunStartIndex > startIndex)
+                {
+                    fullRunList.Add(
+                        new CompositeLicenseExpressionRun(licenseExpression.Substring(startIndex, currentRunStartIndex - startIndex),
+                        CompositeLicenseExpressionRunType.Other));
+                }
+                fullRunList.Add(run);
+                startIndex = currentRunStartIndex + run.Value.Length;
+            }
+
+            if (startIndex < licenseExpression.Length)
+            {
+                fullRunList.Add(
+                    new CompositeLicenseExpressionRun(licenseExpression.Substring(startIndex),
+                    CompositeLicenseExpressionRunType.Other));
+            }
+
+            return fullRunList;
+        }
+
+        private static void InOrderTraversal(NuGetLicenseExpression root, List<CompositeLicenseExpressionRun> runList)
```

Comment:
It's a private method, we shouldn't care too much about how it is called. I prefer to pass the reference to a final list around instead of wrapping pretty much every element into a list and then re-wrapping into other lists as we go up the stack.